remove-data/src/main/java/eu/europeana/metis/remove/discover/ExecutionPluginForest.java

```diff
@@ -0,0 +1,196 @@
+          final ExecutionPluginNode node = new ExecutionPluginNode(execution, plugin);
+          nodesById.put(node.getId(), node);
+        }
+      }
+    }
+
+    // Add all nodes to the lookup map for predecessors
+    final Function<ExecutionPluginNode, PluginLinkByTypeAndTimestamp> createPredecessorKey = node -> new PluginLinkByTypeAndTimestamp(
+        node.getPlugin().getPluginType(), node.getPlugin().getStartedDate());
+    this.nodesByTypeAndTimestamp = nodesById.values().stream()
+        .filter(node -> node.getPlugin().getStartedDate() != null)
+        .collect(Collectors.toMap(createPredecessorKey, Function.identity()));
+
+    // Link parents and children (predecessors and successors)
+    for (ExecutionPluginNode node : nodesById.values()) {
+      final ExecutionPluginNode predecessor = findPredecessor(node);
+      if (predecessor != null) {
+        predecessor.addChild(node);
+      }
+    }
+
+    // Compose the list of root nodes
+    rootNodes = nodesById.values().stream().filter(node -> node.getParent() == null)
+        .collect(Collectors.toList());
+
+    // Find all leaf nodes (candidates for removal)
+    leafNodeIds = nodesById.values().stream()
+        .filter(node -> node.getPlugin().getPluginType() != PluginType.LINK_CHECKING)
+        .filter(ExecutionPluginNode::hasOnlyLinkCheckingDescendants)
+        .map(ExecutionPluginNode::getId).collect(Collectors.toSet());
+  }
+
+  int getNodeCount() {
+    return nodesById.size();
+  }
+
+  int getTreeCount() {
+    return rootNodes.size();
+  }
+
+  int getLeafCount() {
+    return leafNodeIds.size();
+  }
+
+  private ExecutionPluginNode findPredecessor(ExecutionPluginNode node) {
+
+    // Analize the previous plugin information given in the plugin metadata.
+    final Date previousPluginDate = node.getPlugin().getPluginMetadata()
+        .getRevisionTimestampPreviousPlugin();
+    final PluginType previousPluginType = PluginType.getPluginTypeFromEnumName(
+        node.getPlugin().getPluginMetadata().getRevisionNamePreviousPlugin());
+    final ExecutionPluginNode previous;
+    if (previousPluginDate != null && previousPluginType != null) {
+
+      // If a predecessor is named, find it.
+      previous = nodesByTypeAndTimestamp.get(new PluginLinkByTypeAndTimestamp(previousPluginType,
+          previousPluginDate));
+      if (previous == null) {
+        final String message = String
+            .format("Problem with plugin %s: cannot find predecessor with date %s and type %s.",
+                node.getId(), previousPluginDate, previousPluginType);
+        throw new IllegalStateException(message);
+      }
+    } else if (previousPluginDate != null || previousPluginType != null) {
+
+      // If one of them is present, we have an inconsistency.
+      throw new IllegalStateException("Problem with plugin " + node.getId()
+          + ": either the previous plugin type or the previous plugin date is null (but not both).");
+    } else {
+
```

Comment:
As the previous plugin information is only set when the plugin starts executing, this case is meant to cover plugins that have not started executing (because they are in queue or because they are cancelled due to predecessors being cancelled). We don't want to ignore them here, because they may be important for deciding whether or not we can remove a plugin/task.

I added some comments to explain.